<h1> S30 - Evidencia de Aprendizaje Manipulación de Datos con Pandas.</h1>
<h3> Sergio Andres Rios Gomez</h3> 

In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd

<h5> 1. Dataframe de frutas 1 </h5>
<p> Creación de diccionario con los datos, luego se crea el DataFrame "frutas" desde un diccionario. se nombra "frutas" un objeto con la información del DF.</p>

In [2]:
datos = {"Granadilla": [20], "Tomates": [50]}
frutas = pd.DataFrame(datos, index=[0])
frutas

Granadilla  Tomates
0          20       50

<h5> 2. Dataframe de frutas 2 </h5>
<p> Creaciónn de diccionario, pero para este caso, el valor va una lista con los registros, luego se crea una instancia de pandas llamada "ventas_frutas", y recibe como parametros el diccionario y para este caso los valores de los indices.  </p>

In [3]:
datos = {
    "Granadilla": [20, 49],
    "Tomates": [50, 100]
}
ventas_frutas = pd.DataFrame(datos, index=["ventas 2021", "ventas 2022"])
ventas_frutas

Granadilla  Tomates
ventas 2021          20       50
ventas 2022          49      100

<h5> 3. variable utensilios. </h5>
<p> Creación de un diccionario con los datos, luego la Serie "utensilios" desde un diccionario, se especifica el parametro "index" para que quede el nombre de los utensilios.</p> 

In [4]:
# Creando un diccionario con los datos
datos = {
    "Cuchara": [3, "unidades"],
    "Tenedor": [2, "unidades"],
    "Cuchillo": [4, "unidades"],
    "Plato": [5, "unidades"]
}

# Creando la Serie "utensilios" desde un diccionario
utensilios = pd.Series(datos, name="Cocina")

print(utensilios.to_string(index=True))

Cuchara     [3, unidades]
Tenedor     [2, unidades]
Cuchillo    [4, unidades]
Plato       [5, unidades]


<h5> 4. Descarga el dataset 'wine review' desde https://www.kaggle.com/datasets/zynicide/wine-reviews/data Kaggle y cárgalo en un DataFrame llamado review  </h5>
<p> Se nombra un objeto con la ruta en la cual esta el archivo .CSV en el equipo de computo, luego con 'read_csv' de pyyhon se hace la lectura.</p>

In [5]:
file_path = "C:\\Users\\chech\\PC Febrero 2023\\IU Digital\\2024-1\\Analitica de datos\\winemag-data-130k-v2.csv"

# Read the CSV file into a Pandas DataFrame
review = pd.read_csv(file_path)

review

Unnamed: 0   country  \
0                0     Italy   
1                1  Portugal   
2                2        US   
3                3        US   
4                4        US   
...            ...       ...   
129966      129966   Germany   
129967      129967        US   
129968      129968    France   
129969      129969    France   
129970      129970    France   

                                              description  \
0       Aromas include tropical fruit, broom, brimston...   
1       This is ripe and fruity, a wine that is smooth...   
2       Tart and snappy, the flavors of lime flesh and...   
3       Pineapple rind, lemon pith and orange blossom ...   
4       Much like the regular bottling from 2012, this...   
...                                                   ...   
129966  Notes of honeysuckle and cantaloupe sweeten th...   
129967  Citation is given as much as a decade of bottl...   
129968  Well-drained gravel soil gives this wine its c...   
129969  A dry style of Pinot Gris, this is crisp with ...   
129970  Big, rich and off-dry, this is powered by inte...   

                                   designation  points  price  \
0                                 Vulkà Bianco      87    NaN   
1                                     Avidagos      87   15.0   
2                                          NaN      87   14.0   
3                         Reserve Late Harvest      87   13.0   
4           Vintner's Reserve Wild Child Block      87   65.0   
...                                        ...     ...    ...   
129966  Brauneberger Juffer-Sonnenuhr Spätlese      90   28.0   
129967                                     NaN      90   75.0   
129968                                   Kritt      90   30.0   
129969                                     NaN      90   32.0   
129970           Lieu-dit Harth Cuvée Caroline      90   21.0   

                 province             region_1           region_2  \
0       Sicily & Sardinia                 Etna                NaN   
1                   Douro                  NaN                NaN   
2                  Oregon    Willamette Valley  Willamette Valley   
3                Michigan  Lake Michigan Shore                NaN   
4                  Oregon    Willamette Valley  Willamette Valley   
...                   ...                  ...                ...   
129966              Mosel                  NaN                NaN   
129967             Oregon               Oregon       Oregon Other   
129968             Alsace               Alsace                NaN   
129969             Alsace               Alsace                NaN   
129970             Alsace               Alsace                NaN   

               taster_name taster_twitter_handle  \
0            Kerin O’Keefe          @kerinokeefe   
1               Roger Voss            @vossroger   
2             Paul Gregutt           @paulgwine    
3       Alexander Peartree                   NaN   
4             Paul Gregutt           @paulgwine    
...                    ...                   ...   
129966  Anna Lee C. Iijima                   NaN   
129967        Paul Gregutt           @paulgwine    
129968          Roger Voss            @vossroger   
129969          Roger Voss            @vossroger   
129970          Roger Voss            @vossroger   

                                                    title         variety  \
0                       Nicosia 2013 Vulkà Bianco  (Etna)     White Blend   
1           Quinta dos Avidagos 2011 Avidagos Red (Douro)  Portuguese Red   
2           Rainstorm 2013 Pinot Gris (Willamette Valley)      Pinot Gris   
3       St. Julian 2013 Reserve Late Harvest Riesling ...        Riesling   
4       Sweet Cheeks 2012 Vintner's Reserve Wild Child...      Pinot Noir   
...                                                   ...             ...   
129966  Dr. H. Thanisch (Erben Müller-Burggraef) 2013 ...        Riesling   
129967                  Citation 2004 P

<h5> 5. Visualiza las primeras filas del DataFrame </h5>
<p> El método head() en Pandas es una función que se utiliza para obtener las primeras n filas de un DataFrame. Es una herramienta útil para visitar rápidamente los datos y hacerse una idea general de su contenido, estructura y formato.</p>

In [6]:
review.head()

Unnamed: 0   country                                        description  \
0           0     Italy  Aromas include tropical fruit, broom, brimston...   
1           1  Portugal  This is ripe and fruity, a wine that is smooth...   
2           2        US  Tart and snappy, the flavors of lime flesh and...   
3           3        US  Pineapple rind, lemon pith and orange blossom ...   
4           4        US  Much like the regular bottling from 2012, this...   

                          designation  points  price           province  \
0                        Vulkà Bianco      87    NaN  Sicily & Sardinia   
1                            Avidagos      87   15.0              Douro   
2                                 NaN      87   14.0             Oregon   
3                Reserve Late Harvest      87   13.0           Michigan   
4  Vintner's Reserve Wild Child Block      87   65.0             Oregon   

              region_1           region_2         taster_name  \
0                 Etna                NaN       Kerin O’Keefe   
1                  NaN                NaN          Roger Voss   
2    Willamette Valley  Willamette Valley        Paul Gregutt   
3  Lake Michigan Shore                NaN  Alexander Peartree   
4    Willamette Valley  Willamette Valley        Paul Gregutt   

  taster_twitter_handle                                              title  \
0          @kerinokeefe                  Nicosia 2013 Vulkà Bianco  (Etna)   
1            @vossroger      Quinta dos Avidagos 2011 Avidagos Red (Douro)   
2           @paulgwine       Rainstorm 2013 Pinot Gris (Willamette Valley)   
3                   NaN  St. Julian 2013 Reserve Late Harvest Riesling ...   
4           @paulgwine   Sweet Cheeks 2012 Vintner's Reserve Wild Child...   

          variety               winery  
0     White Blend              Nicosia  
1  Portuguese Red  Quinta dos Avidagos  
2      Pinot Gris            Rainstorm  
3        Riesling           St. Julian  
4      Pinot Noir         Sweet Cheeks

<h5> 6.Utiliza el método .info() para averiguar cuántas entradas hay. </h5>
<p> El método .info() proporciona un resumen conciso de la estructura y la información del DataFrame. En este caso, nos indica que el DataFrame tiene un índice de rango (RangeIndex) que va desde 0 hasta 129970, lo que implica que hay 129971 filas en total. Un total de 14 columnas y describe el tipo de dato de cada una de las columnas.</p>

In [7]:
review.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 129971 entries, 0 to 129970
Data columns (total 14 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Unnamed: 0             129971 non-null  int64  
 1   country                129908 non-null  object 
 2   description            129971 non-null  object 
 3   designation            92506 non-null   object 
 4   points                 129971 non-null  int64  
 5   price                  120975 non-null  float64
 6   province               129908 non-null  object 
 7   region_1               108724 non-null  object 
 8   region_2               50511 non-null   object 
 9   taster_name            103727 non-null  object 
 10  taster_twitter_handle  98758 non-null   object 
 11  title                  129971 non-null  object 
 12  variety                129970 non-null  object 
 13  winery                 129971 non-null  object 
dtypes: float64(1), int64(2), object(11)


<h5> 7. ¿Cuál es el precio promedio? </h5>
<p> El método mean() de Pandas calcula el promedio de una columna de manera predeterminada, ignorando los valores NaN </p> 

In [11]:
promedio_Price = review['price'].mean()
mean_price = promedio_Price.round(2)
print('El precio promedio de los precios es: ', mean_price)

El precio promedio de los precios es:  35.36


<h5> 8. ¿Cuál es el precio más alto pagado? </h5>
<p> El método max() de Pandas recorre la columna "price" y devuelve el valor más alto. La variable precio_maximo almacenará el precio más alto encontrado en la columna. El método max() ignora los valores NaN por defecto. Si hay valores NaN en la columna "price", no se considerarán para calcular el precio máximo.</p>

In [14]:
precio_maximo = review['price'].max()
print('El precio más alto es: ', precio_maximo)

El precio más alto es:  3300.0


<h5> 9. Crea un daframe con todos los vinos de california: </h5>

In [28]:
vinos_california = review.loc[review['province'] == 'California']
vinos_california

Unnamed: 0 country                                        description  \
10              10      US  Soft, supple plum envelopes an oaky structure ...   
12              12      US  Slightly reduced, this wine offers a chalky, t...   
14              14      US  Building on 150 years and six generations of w...   
23              23      US  This wine from the Geneseo district offers aro...   
25              25      US  Oak and earth intermingle around robust aromas...   
...            ...     ...                                                ...   
129942      129942      US  This is classic in herbaceous aromas and flavo...   
129945      129945      US  Hailing from one of the more popular vineyards...   
129949      129949      US  There's no bones about the use of oak in this ...   
129950      129950      US  This opens with herbaceous dollops of thyme an...   
129952      129952      US  This Zinfandel from the eastern section of Nap...   

                             designation  points  price    province  \
10                        Mountain Cuvée      87   19.0  California   
12                                   NaN      87   34.0  California   
14                                   NaN      87   12.0  California   
23                   Signature Selection      87   22.0  California   
25                   King Ridge Vineyard      87   69.0  California   
...                                  ...     ...    ...         ...   
129942                               NaN      90   35.0  California   
129945  Jurassic Park Vineyard Old Vines      90   20.0  California   
129949                  Barrel Fermented      90   35.0  California   
129950                     Blocks 7 & 22      90   35.0  California   
129952                               NaN      90   22.0  California   

                 region_1       region_2     taster_name  \
10            Napa Valley           Napa  Virginie Boone   
12       Alexander Valley         Sonoma  Virginie Boone   
14          Central Coast  Central Coast   Matt Kettmann   
23            Paso Robles  Central Coast   Matt Kettmann   
25           Sonoma Coast         Sonoma  Virginie Boone   
...                   ...            ...             ...   
129942      Sonoma County         Sonoma  Virginie Boone   
129945  Santa Ynez Valley  Central Coast   Matt Kettmann   
129949        Napa Valley           Napa  Virginie Boone   
129950        Napa Valley           Napa  Virginie Boone   
129952      Chiles Valley           Napa  Virginie Boone   

       taster_twitter_handle  \
10                   @vboone   
12                   @vboone   
14             @mattkettmann   
23             @mattkettmann   
25                   @vboone   
...                      ...   
129942               @vboone   
129945         @mattkettmann   
129949               @vboone   
129950               @vboone   
129952               @vboone   

                                                    title             variety  \
10      Kirkland Signature 2011 Mountain Cuvée Caberne...  Cabernet Sauvignon   
12      Louis M. Martini 2012 Cabernet Sauvignon (Alex...  Cabernet Sauvignon   
14               Mirassou 2012 Chardonnay (Central Coast)          Chardonnay   
23      Bianchi 2011 Signature Selection Merlot (Paso ...              Merlot   
25      Castello di Amorosa 2011 King Ridge Vineyard P...          Pinot Noir   
...                                                   ...                 ...   
129942   Arrowood 2010 Cabernet Sauvignon (Sonoma County)  Cabernet Sauvignon   
129945  Birichino 2013 Jurassic Park Vineyard Old Vine...        Chenin Blanc   
129949  Flora Springs 2013 Barrel Fermented Chardonnay...          Chardonnay   
129950  Hendry 2012 Blocks 7 & 22 Zinfandel (Napa Valley)           Zinfandel   
129952             Houdini 2011 Zinfandel (Chiles Valley)           Zinfandel   

                     winery  
10       Kirkland Signature  
12         Louis M. Martini  
14           

<h5> 10. Usa idxmax() para encontrar el índice del vino con el precio más alto y luego usa loc para obtener toda la información de ese vino específico. </h5>
<p> Este código primero encuentra el índice del vino con el precio más alto en el DataFrame "vinos_california" utilizando el método idxmax() aplicado a la columna "price". Luego, utiliza la función loc[] para obtener toda la información del vino en ese índice específico.</p>
<p> El método idxmax() en Pandas se utiliza para encontrar el índice de la fila con el valor máximo en una columna o eje específico de un DataFrame.</p>
<p> DataFrame.idxmax ( eje = 0 , skipna = True , numeric_only = False ). Índice de retorno de la primera ocurrencia del máximo sobre el eje solicitado. Se excluyen los valores NA/nulos. los Parámetros descritos en la documentación son opcionales </p>

In [30]:
print("El vino más caro de california es:")
indice_vino_maximo_precio = vinos_california['price'].idxmax()

# Obtener toda la información del vino con el precio más alto
vino_mas_caro = vinos_california.loc[indice_vino_maximo_precio]

print(vino_mas_caro)

El vino más caro de california es:
Unnamed: 0                                                          120391
country                                                                 US
description              The nose on this single-vineyard wine from a s...
designation                                            Roger Rose Vineyard
points                                                                  91
price                                                               2013.0
province                                                        California
region_1                                                       Arroyo Seco
region_2                                                     Central Coast
taster_name                                                  Matt Kettmann
taster_twitter_handle                                        @mattkettmann
title                    Blair 2013 Roger Rose Vineyard Chardonnay (Arr...
variety                                                         C

<h5> 11. Cuáles son los tipos de uva más comunes en California? </h5>

In [41]:
# Crear un DataFrame con variedad de vino y su frecuencia
dataframe_frecuencia_variedad = vinos_california['variety'].value_counts().reset_index()

# Renombrar columnas
dataframe_frecuencia_variedad.columns = ['variedad_vino', 'frecuencia']


dataframe_frecuencia_variedad

variedad_vino  frecuencia
0                        Pinot Noir        6896
1                Cabernet Sauvignon        5693
2                        Chardonnay        5183
3                         Zinfandel        2639
4                             Syrah        1870
..                              ...         ...
189  Tempranillo-Cabernet Sauvignon           1
190         Syrah-Grenache-Viognier           1
191                      Negroamaro           1
192                  Moscato Giallo           1
193                       Alvarinho           1

[194 rows x 2 columns]

<p> Identificar las variedades de vinos que tienen una frecuencia mayor a 100, es decir que dentro de los vinos de California están en más de 100 registros.  </p>

In [44]:
vinos_frecuencia_mayor_100 = dataframe_frecuencia_variedad[dataframe_frecuencia_variedad['frecuencia'] > 100]
variedades_con_alta_frecuencia = vinos_frecuencia_mayor_100['variedad_vino'].tolist()
print(variedades_con_alta_frecuencia)

['Pinot Noir', 'Cabernet Sauvignon', 'Chardonnay', 'Zinfandel', 'Syrah', 'Sauvignon Blanc', 'Red Blend', 'Merlot', 'Bordeaux-style Red Blend', 'Petite Sirah', 'Sparkling Blend', 'Rosé', 'Rhône-style Red Blend', 'Viognier', 'Grenache', 'White Blend', 'Cabernet Franc', 'Pinot Grigio', 'Riesling', 'Sangiovese', 'Malbec', 'Barbera', 'Pinot Gris', 'Meritage', 'Rhône-style White Blend', 'Tempranillo', 'Gewürztraminer', 'Mourvèdre', 'G-S-M', 'Petit Verdot', 'Albariño']


<h5> 12. Cuáles son los 10 tipos de uva más comunes en California? </h5>

In [40]:
# Crear un DataFrame con variedad de vino y su frecuencia
dataframe_frecuencia_variedad = vinos_california['variety'].value_counts().reset_index()

# Renombrar columnas
dataframe_frecuencia_variedad.columns = ['variedad_vino', 'frecuencia']


dataframe_frecuencia_variedad.head(10)

variedad_vino  frecuencia
0                Pinot Noir        6896
1        Cabernet Sauvignon        5693
2                Chardonnay        5183
3                 Zinfandel        2639
4                     Syrah        1870
5           Sauvignon Blanc        1807
6                 Red Blend        1804
7                    Merlot        1391
8  Bordeaux-style Red Blend         876
9              Petite Sirah         694